### DISCUSSION TITLE
请问 gent 是否支持设置 client 发送时自身的 ip 和 port ？

### DISCUSSION THREAD

**Discussion by lwwgo** [panjf2000/gnet#416] [Q&A]
1、gent 版本：v2.1.2 或者其他有client的版本
2、指定发送client自身的ip 和 port 主要用在NAT场景下，或者能同时用server 和 client，client能够指定server使用的ip 和 port，这样server 通过nat 与外部打洞成功后，client可以使用server的ip和端口号。
不知道目前是否支持？

**Comment by panjf2000** [comment]
不太理解为什么要手动指定，通常不是在 NAT gateway 那里就可以自动把 IP 替换掉吗？

**Reply by lwwgo** [reply]
举例：
10.1.1.1:1000 和 10.2.2.2:2000 在两个NAT 后，10.2.2.2:2000 和 10.1.1.1:1000 不能直接通信，需要先打洞（https://rebootcat.com/2021/03/28/p2p_nat_traversal/），打洞完成后 10.1.1.1:1000 与 10.2.2.2:2000 可以互相通信了，但是若 此时再启动一个gnet client向10.1.1.1:1000 发送数据，因为client的ip和prot无法指定为 已经打洞成功的 ip:port(10.2.2.2:2000)，所以client发出的消息，对方的nat会拦截下，无法发送到  10.1.1.1:1000 。
因此，若client可以指定发送数据时候的ip(10.2.2.2) 和 port(2000)，则能够解决这种场景下的问题。

**Reply by panjf2000** [reply]
目前 gnet client 并不支持设置 ip/port，要支持你的场景，需要开发新特性，我会评估一下。

**Reply by lwwgo** [reply]
> 目前 gnet client 并不支持设置 ip/port，要支持你的场景，需要开发新特性，我会评估一下。

好的，感谢潘少！若支持了，通知哦

**Reply by panjf2000** [reply]
今天又看了下相关的代码，我觉得你可以这么做：
```go
// 先用标准库 net 指定 local address
c, err := net.DialTCP(network string, laddr, raddr *TCPAddr)
// 然后再调用 gnet client 的 Enroll() 方法，内部使用 dup 复制 fd，保留原 ip 和 port.
gnetConn, err := gnetClient.Enroll(c)
```

**Reply by lwwgo** [reply]
> 今天又看了下相关的代码，我觉得你可以这么做：
> 
> ```go
> // 先用标准库 net 指定 local address
> c, err := net.DialTCP(network string, laddr, raddr *TCPAddr)
> // 然后再调用 gnet client 的 Enroll() 方法，内部使用 dup 复制 fd，保留原 ip 和 port.
> gnetConn, err := gnetClient.Enroll(c)
> ```

有个fork这样做的，在udp下可用，https://github.com/Ricardo-Evans/gnet/blob/master/client_unix.go
